File: 0590-n-ary-tree-postorder-traversal/iterative-depth-first-search/cpp/solution.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> children;

    Node() {}

    Node(int _val) {
        val = _val;
    }

    Node(int _val, vector<Node*> _children) {
        val = _val;
        children = _children;
    }
};
*/

class Solution {
public:
    vector<int> postorder(Node* root) {
        vector<int> result;
        stack<Node*> stk;
        stk.push(root);

        while (!stk.empty()) {
            Node* node = stk.top();
            stk.pop();

            if (node) {
                result.push_back(node->val);
                // Push children left-to-right so that rightmost child stays on top and pops first
                for (Node* child : node->children) {
                    if (child) {
                        stk.push(child);
                    }
                }
            }
        }

        // Reverse to convert [Root -> Rightmost ... Leftmost] into  [Leftmost ... Rightmost -> Root] (i.e. postorder)
        reverse(result.begin(), result.end());
        return result;
    }
};

File: 0590-n-ary-tree-postorder-traversal/iterative-depth-first-search/java/solution.java
/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> children;

    public Node() {}

    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, List<Node> _children) {
        val = _val;
        children = _children;
    }
}
*/

class Solution {
    public List<Integer> postorder(Node root) {
        List<Integer> result = new ArrayList<>();
        if (root == null) return result;
        
        Deque<Node> stack = new ArrayDeque<>();
        stack.push(root);

        while (!stack.isEmpty()) {
            Node node = stack.pop();

            if (node != null) {
                result.add(node.val);
                if (node.children != null) {
                    // Push children left-to-right so that rightmost child stays on top and pops first
                    for (Node child : node.children) {
                        if (child != null)  stack.push(child);
                    }
                }
            }
        }

        // Reverse to convert [Root -> Rightmost ... Leftmost] into [Leftmost ... Rightmost -> Root] (i.e. postorder)
        Collections.reverse(result);
        return result;
    }
}
